### ISSUE TITLE
DeploymentManifestCleaner timing out with large multi-tenancy setup

### ISSUE THREAD

**Issue by thebigmoosey** [OctopusDeploy/Issues#3531]
There's a blocking performance issue with the `DeploymentManifestCleaner` [here](https://github.com/OctopusDeploy/OctopusDeploy/blob/master/source/Octopus.Core/Model/Projects/DeploymentManifestCleaner/DeploymentManifestCleaner.cs#L59) where we run a sub-query against the `MultiTenancyDashboard` view. This is causing the Retention policy task to timeout/fail. This particular case has a relatively large number of tenants (> 100).

See the `2017-05-29.bak` file in SmartFile [here](https://octopusdeploy.smartfile.com/ftp/private/browse/#path=/Mark%20Siedle%20support%20uploads&page=1&limit=100) for an example database where this query runs for a very very long time. Note: I was not able to reproduce this with my standard MT installation (trucks sample), so asked customer for an export and confirmed against their data.

The query runs fine until we add the last condition: `AND Deployment.Id NOT IN (SELECT Id FROM MultiTenancyDashboard)`

## Possible solution to consider
We could get the results of the dashboard view query and pass them in as a parameter to the subquery (when I did this manually in SQL it ran fast).

Or we could split this into two separate queries and run the deployment exclusion logic in-memory on the server, rather than trying to do this all in the DB (where I'm guessing the subquery is locking because of something from the `MultiTenancyDashboard` view).

## Index recommendation
Query analyser says we should also consider an index on the VariableSet table's `OwnerId` column. Would be worth doing as part of this fix.

Source: http://help.octopusdeploy.com/discussions/problems/54817

**Comment by droyad** [comment]
With the first solution (pass to subquery), we would still be taking out a readlock on the dashboard.

If we run the select query outside of the transaction, we may end up deleting deployments that have come back onto the dashboard while the loop was being run. Then again, I can't think of when that'd happen.

I believe we originally kept the read of variable sets as a steam instead of in-memory was that the list can get quite large and customers were having performance problems.

I think the best option is to do the read in it's own transaction, but set the isolation level and/or table hints so that no locks are taken out. Then open a transaction and commit it within the foreach loop. A little bit more overhead, but each delete/update pair can be considered independent. Then we also don't run up the transaction size.

**Comment by droyad** [comment]
It looks like I misunderstood the issue, I thought that we were getting blocked by a lock. The query still ran slow even when run in isolation in SSMS. Rebuilding indexes and/or updating statistics resolves the problem. It must have been a REALLY bad query plan as it went from over 18 minutes to < 1 sec.

**Comment by droyad** [comment]
The problem reoccurs after a short time. We should do the above modifications and provide some query hints

**Comment by octoreleasebot** [comment]
Release Note: Performance improvements to the deployment manifest clean up queries
